src/impl_owned_array.rs

```diff
@@ -59,3 +69,558 @@ where
+    /// along the "growing axis".
+    ///
+    /// Ensure appending is efficient by for example starting appending an empty array and
+    /// always appending along the same axis. For rows, ndarray's default layout is efficient for
+    /// appending.
+    ///
+    /// Notice that an empty array (where it has an axis of length zero) is the simplest starting
+    /// point. The amortized average complexity of the append is O(m) where *m* is the length of
+    /// the row.
+    ///
+    /// ```rust
+    /// use ndarray::{Array, ArrayView, array};
+    ///
+    /// // create an empty array and append
+    /// let mut a = Array::zeros((0, 4));
+    /// a.append_row(ArrayView::from(&[ 1.,  2.,  3.,  4.])).unwrap();
+    /// a.append_row(ArrayView::from(&[-1., -2., -3., -4.])).unwrap();
+    ///
+    /// assert_eq!(
+    ///     a,
+    ///     array![[ 1.,  2.,  3.,  4.],
+    ///            [-1., -2., -3., -4.]]);
+    /// ```
+    pub fn append_row(&mut self, row: ArrayView<A, Ix1>) -> Result<(), ShapeError>
+    where
+        A: Clone,
+    {
+        self.append(Axis(0), row.insert_axis(Axis(0)))
+    }
+
+    /// Append a column to an array
+    ///
+    /// ***Errors*** with a shape error if the length of the column does not match the length of
+    /// the columns in the array.<br>
+    ///
+    /// The memory layout of the `self` array matters for ensuring that the append is efficient.
+    /// Appending automatically changes memory layout of the array so that it is appended to
+    /// along the "growing axis".
+    ///
+    /// Ensure appending is efficient by for example starting appending an empty array and
+    /// always appending along the same axis. For columns, column major ("F") memory layout is
+    /// efficient for appending.
+    ///
+    /// Notice that an empty array (where it has an axis of length zero) is the simplest starting
+    /// point. The amortized average complexity of the append is O(m) where *m* is the length of
+    /// the row.
+    ///
+    /// ```rust
+    /// use ndarray::{Array, ArrayView, array};
+    ///
+    /// // create an empty array and append
+    /// let mut a = Array::zeros((2, 0));
+    /// a.append_column(ArrayView::from(&[1., 2.])).unwrap();
+    /// a.append_column(ArrayView::from(&[-1., -2.])).unwrap();
+    ///
+    /// assert_eq!(
+    ///     a,
+    ///     array![[1., -1.],
+    ///            [2., -2.]]);
+    /// ```
+    pub fn append_column(&mut self, column: ArrayView<A, Ix1>) -> Result<(), ShapeError>
+    where
+        A: Clone,
+    {
+        self.append(Axis(1), column.insert_axis(Axis(1)))
+    }
+}
+
+impl<A, D> Array<A, D>
+    where D: Dimension
+{
+    /// Move all elements from self into `new_array`, which must be of the same shape but
+    /// can have a different memory layout. The destination is overwritten completely.
+    ///
+    /// ***Panics*** if the shapes don't agree.
+    pub fn move_into(mut self, new_array: ArrayViewMut<MaybeUninit<A>, D>) {
```

Comment:
I just realized that we could change this signature to

```rust
    pub fn move_into<S2>(mut self, new_array: &mut ArrayBase<S2, D>) {
    where
        S2: AssignElem<A>,
```

which would make it more useful. The primary issue is dealing with panics, especially since `AssignElem` is a safe trait. It's tricky to deal with panics in the implementation without leaking elements on panic or aborting on panic. Even if we used a new `unsafe` trait instead of `AssignElem`, the overwritten elements can panic when they're dropped. I suppose an easy solution would be to move the elements being overwritten into a temporary `Vec`, and then drop the temporary `Vec`, but that seems pretty inefficient.

Reply:
Re ArrayViewMut vs &mut ArrayBase: view mut is closer to what we actually need (we convert to view mut unconditionally in the Zip); so "zero-cost" interface would be something like `Into<ArrayViewMut<_>>`. It's only an ergonomic difference and not functional, I believe.

It's right that we only use MaybeUninit here, so we only overwrite, so that we can never have panics in that critical section. I would put supporting anything more as something to be added later.  It is intended that we add a method that makes it easy to convert a view mut of T into a view mut of MaybeUninit<T>, and use this method broadly that way.

Reply:
It can feel like we're not making the 100% solution here and maybe that's ok. We grow features bit by bit.